### ISSUE TITLE
[Epic] Simplify CRUD Queries v1

### ISSUE THREAD

**Issue by ajinkyakulkarni** [appsmithorg/appsmith#4559]
# Objective
Reduce the time to ship useful plugins to 1 week. 
Today it takes 8 weeks to get plugins into a usable state. While the initial plugin does come out fairly quickly, it usually requires a lot of iteration to become useful. There are 2 core gaps here
- Lack of clarity on what features should the plugin support
- Poor UX on the plugin which makes it unusable without documentation

## Key Points
- Standardise Plugin Requirements for all CRUD operations
- A design system and guidelines so we can add high-quality integrations with as little design as possible.
- Components for complex interactions like array object input.
- A solid contributor's guide to integrations by September 2021, so that we can scale integrations during Hacktoberfest.

# Success Metrics
| Goal  | Metric |
| ------------- | ------------- |
| High-Quality DB Integrations | 15 DB Integrations that anyone can use without documentation  |
| High-Quality SAAS Integrations | 10 SAAS Integrations that anyone can use without documentation |
| Simple Contribution Platform | Get 5 external contributions |

# Requirements
| Requirement   | User Story | Importance | Notes |
| ------------- | ------------- | ------------- | ------------- |
| Unified Interface Generator | Our system should be able to generate the interface for database/SAAS integrations with minimal input from the contributor | High |
| Intuitive Interface | A user should have a homogenous intuitive experience using integrations on appsmith | High |
| Homogenous DB Functionality | A user should be able to achieve the same operations across DB integrations  | High |
| Raw Mode | A user should be able to access the underlying query construct and make use of it with the aid of documentation. | High |
| SAAS Ingestion App | A contributor should be easily able to add a SAAS integration without help from the appsmith team | High  |
| DB Contribution Interface | A user should have a simple interface with documentation to contribute a new database integration | Medium |

# Out of Scope
ORM Method of querying data

# Unified Interface Solution
Databases are generally structured with homogenous entities i.e Tables, Collections, Files, etc and these are the entities on top of which operations are generally performed. DB Interfaces can be modeled around these entities to simplify the querying experience for users. DBs are also largely used for CRUD and thus the number of operations to be performed on them can also be limited to Create, Read, Update & Delete. For our existing databases, this could look like
1. Users must select a DB (Google Sheet, Firebase, SQL, Mongo, S3, Elastic, Redis, Dynamo, Big Query)
2. A user must select an operation
    - Fetch
    - Insert
    - Update (Not valid for S3)
    - Delete
    - Aggregate (Not valid for S3)
    - Bulk Insert
    - Bulk Update
    - Bulk Delete
    - Raw Query (Not valid for Firebase / S3)
    - Custom Action (?)
3. A user must select an entity / sub-entity (optional) / ID (optional) in the DB. They can be configured Dynamically as well.
    - Google Sheets (SpreadSheet)
      - Sheet Name (Optional)
        - Table Header Index
          - Row Index 
    - SQL (Table)
      - Primary Key 
    - Mongo/Firestore/Firebase (Collection)
      - Object ID (without object Id syntax) 
    - Redis (DB)
      - Key
    - S3 (Bucket)
      - File (Optional) (Only valid for Fetch Op)
    - Elastic (Index)
      - Document (Optional)
    - Big Query (Projects)
      - Datasets (Optional)
        - Models (Optional)
        - Routines (Optional)
        - Tables (Optional) 
      - Jobs (Optional)
      - Queries (Optional)
    - SnowFlake (Table)

## Fetch
This operation is used to fetch the data inside the entity. This operation requires a user to configure
1. Where Clause: This is a UI element that allows the user to configure multiple conditions to fetch the data. The values can be configured dynamically. if the plugin does not support this, the server should do the filtering.
2. Limit: This is a UI element that allows a user to configure the number of rows of the entity to return. If the plugin does not support this, the server should limit the amount of data returned. A default value must be present for every plugin to prevent excessive data from being returned.
3. Offset: This is a UI element that allows users to configure how many records to skip from the first record that matches the where clause. The default value of this field should be 0.
4. OrderBy: This is a UI element that selects the column(s) by which to order the data. If the plugin does not support filtering server-side, the server should order the data. It should also allow users to select an ascending / descending order
5. Projection: This is a UI element that selects which fields of the response to return. This field can be hidden for single key responses like S3 files.
6. Options: This is a UI section that contains plugin-specific additional parameters such as
    - Enable URL Signing (S3)
    - URL Expiry Duration (S3)

The response of this operation should always be an array of objects with keys and values.

## Insert
This operation is used to insert a new row into the entity. This operation requires a user to enter an object. The user should be made aware of the object format and have a sample object to refer to.

The response of this operation should be a JSON of the form
```
{
"message": "Row successfully Inserted"
}
```

## Update
This operation is used to update a row in the entity. This operation requires a user to enter 3 fields
1. New Object properties: 
2. Update Diffs: This switch will update only the keys specified for JSON objects. This is enabled by default. If this is disabled, unspecified keys will be deleted from the object. This field should have a tooltip explaining how it works

The response of this operation should be a JSON of the form
```
{
"message": "Row successfully updated"
}
```

## Delete
This operation is used to delete a row in the entity. This operation requires a user to enter 3 fields
```
{
"message": "Row successfully deleted"
}
```

## Bulk Insert
This operation is used to insert a new row into the entity. This operation requires a user to enter an array of objects. The user should be made aware of the object format and have a sample object to refer to.
The response of this operation should be a JSON of the form
```
{
"message": "10 Rows successfully Inserted"
}
```

## Bulk Update
This operation is used to update multiple rows in the entity. This operation requires a user to enter the following fields
1. Update Type:
    - Where Update: 
      - Filter: This section should contain the where / limit / offset / order by sections in the Fetch OP.
      - New Object properties: This operation requires a user to enter an object. The user should be made aware of the object format and have a sample object to refer to.
    - Objects Update: This operation requires a user to enter an array of objects. The user should be made aware of the object format and have a sample object to refer to. Each object of the array has an Id field.
2. Update Differences: This property will update only the keys specified for JSON objects. This is enabled by default. If this is disabled, unspecified keys will be deleted from the object.

The response of this operation should be a JSON of the form
```
{
"message": "10 Rows successfully updated"
}
```

## Bulk Delete
This operation is used to delete multiple rows in the entity. This operation requires a user to enter 2 fields.
- Filter: This section should contain the where / limit / offset /order by sections in the Fetch OP.
```
{
"message": "10 Rows successfully deleted"
}
```

## Aggregate
This operation is used to return aggregate data of results. The backend will do the aggregation logic for plugins that do not support it.
1. Operator: The user can select from the following operations AVG | MIN | MAX | COUNT | DISTINCT | SUM
2. Column: The user can select a key of the entity to apply the operation on 
3. GroupBy: The user can select a key of the entity to group the response operation by

The response of this operation should always be an array of objects with keys and values.

## Design Components Required
1. Operation Selector
2. Entity / Sub Entity / ID Selector
2. Where Filtering Section
3. Insert / Update Body
4. Bulk Insert / Update Body
5. Projection Selector
6. Options

# DB Contribution Solution

# SAAS Ingestion Solution
Steps required for a new integration are recorded [here](https://www.notion.so/appsmith/Contributor-guidelines-c6d7b88e059f487c8a7f073777e821fc).

# Figma
https://www.figma.com/file/5x2k7Uz4TKHRqKPtdqgsBB/Unified-Query-Interface?node-id=118%3A2750

**Comment by areyabhishek** [comment]
I have a recommendation here because it seems like we'll now be using forms at a lot of places. If we are using forms, always expose the entire form data as a JSON so that it's easy to copy paste across queries/apps/orgs, etc. The advantage of form is that you can input things quickly, but they make copy/paste difficult. So let's make it a default option to always be able to access the JSON. It's a small feature that'll rapidly improve the DX.

**Comment by ajinkyakulkarni** [comment]
Internal meeting notes https://www.notion.so/appsmith/UQI-Meeting-Notes-July-16-2021-00413114fa064439af4be6310d42a5b6

**Comment by Nikhil-Nandagopal** [comment]
Next Action Items
- Firestore UQI Migration @sumitsum 
- Type Validations for fields @rashmi-rai 
- Push where clause & entity selector @trishaanand 
- Publish 1 plugin @nidhi-nair
- S3 content type bug for file upload @nidhi-nair

**Comment by ajinkyakulkarni** [comment]
**Meeting notes**
- Firestore UQI migration will be done by Friday - Sumit
- Type validation for field may slip over next week - Rashmi
- Still working on where clause & entity selector (aim to finish next week) - Trisha
- Need help with merging where clause frontend changes to release - Rashmi
- Then start work on backend changes  - Trisha
- Airtable plugin is under testing (Prapulla has raised bugs) - Nidhi
- S3 content type bug for file upload is on release (done) - Nidhi
- Parth needs input on https://www.notion.so/appsmith/Request-Body-Component-UQI-d24ce5f6092645e9afbd4439ab44adda
- Next week: fix SaaS plugin bugs (better error handling, OAuth) - Nidhi
- Rashmi will continue helping on UQI until the work in the new pod picks up
- Enough work is left in UQI project for the next 3 weeks

**Comment by trishaanand** [comment]
Meeting notes :
- Firestore UQI migration will be finished in a day or two - Sumit
- Merge where clause & entity selector to release (dev is complete) - Trisha
- Backend implementation for projection, pagination & sorting to begin post 24th - Sumit
- Field type validation is being worked on - Rashmi
- Entity Selector front end components being worked on - Ayush
- Onboarding for the project - Ayush

To do : Team should get together and revisit the timelines for delivering complete UQI for GSheets, S3, Firestore and Mongo

**Comment by sumitsum** [comment]
Meeting Notes:
- Sumit - backend implementation for Projection, Pagination and Sorting 
- Ayush - working on Entity Selector, onboarding new members - Leo, Aman (?)
- Rashmi - working on Field type validation
- Nidhi - Port GSheet to UQI schema
- Ade - pick up tasks post consultation with Ayush
- Favor - working on OAuth front end
- Ajinkya, Ayush - need to check with Parth regarding design for Sort feature
- Team will get together and estimate timeline for delivering complete UQI for GSheets, S3, Firestore and Mongo

**Comment by nidhi-nair** [comment]
Meeting Notes: 

- [ ] SaaS OAuth2 configs client side development is done, the changes need to be merged in (Favour)
- [ ] Testing OAuth2 has been approximated to take two more weeks based on the above (Aman and Nidhi)
- [ ] Google sheets UQI needs client side support for the new form type, ETA end of week (Ade)
- [ ] Google sheets UQI will then need to be migrated, ETA end of next week (Nidhi)
- [ ] Firestore UQI PR needed some changes to avoid breaks in the DB migrations, that is WIP (Sumit)
- [ ] Firestore also needs some client side changes that is yet to be picked up (estimated to be minor fixes)
- [ ] S3 get structure has test failures, one of them needs to be removed and others are getting fixed (Sumit)
- [ ] SaaS app bugs are being worked on this week, aiming to clear Airtable through as a result (Aman and Nidhi)
- [ ] Pending PRs need to be closed, we need to redistribute these PRs and actively work on closing them (Aman, Rashmi, Ade)
- [ ] For form mode - native mode, we have decided to not support conversion from native to form, and push form to native as an enhancement outside of this project. UQI will simply introduce two separate sets of data for each mode that do not interact with each other. Estimation for the feature is pending (Ajinkya and Rohan)
- [ ] Sorting components designs are ready. https://www.figma.com/file/Bp5IgfLhYAPhGhRkgKHuR2/?node-id=215%3A15403
- [ ] Request body component designs will be looked over again for a final sign off (Rohan)

**Comment by nidhi-nair** [comment]
Meeting Notes:

- [ ] UQI components client side implementation will be done by next week
- [ ] Validations might need some more inputs, Ayush to talk to Rashmi about this and get back
- [ ] Firestore plugin changes should be merged this week
- [ ] S3 PR needs to be looked into, Sumit will be closing it this week as well
- [ ] Google sheets component implementation will start as the components get implemented. The migration is blocked by adding support for native and appsmith modes
- [ ] Airtable is ready for QA again, this should be on prod by next week
- [ ] Saas app is currently bug free. We'll be copying this app into release and connecting the current one to prod. Deployable version will be ready after the git sync issue is fixed)
- [ ] Need to create Github issues for all CRUD operations that are missing for plugins
- [ ] Need to take a look at whether the component implementation by next week will take care of all CRUD operations
- [ ] Need to create issues to track out of scope items for later: Field wise error handling

**Comment by rohan-arthur** [comment]
wireframes: 
https://www.notion.so/appsmith/Pagination-Component-UQI-9872b0252a254b64b9af6b855792e1a0

**Comment by rohan-arthur** [comment]
_Copying Nidhi's comment here and adding a few more points_

Meeting Notes:

 

- [ ] UQI components client side implementation will be done by next week
- [ ] Validations might need some more inputs, Ayush to talk to Rashmi about this and get back
- [ ] Firestore plugin changes should be merged this week
- [ ] S3 PR cypress test failures needs to be looked into - Sumit
- [ ] Google sheets component implementation will start as the components get implemented. The migration is blocked by adding support for native and appsmith modes
- [ ] Airtable is ready for QA again, this should be on prod by next week
- [ ] Saas app is currently bug free. We'll be copying this app into release and connecting the current one to prod. Deployable version will be ready after the git sync issue is fixed)
- [ ] Need to create Github issues for all CRUD operations that are missing for plugins
- [ ] Need to take a look at whether the component implementation by next week will take care of all CRUD operations
- [ ] Need to create issues to track out of scope items for later: Field wise error handling
- [ ] PR to add server side support for Pagination, Sorting and Projection is on review cycle and should be closed by this week end
- [ ] Add pagination, sorting, projection components to plugins
- [ ] MongoDB form to native query conversion
- [ ] Decide on [design for request body components](https://www.figma.com/file/Bp5IgfLhYAPhGhRkgKHuR2/UQI?node-id=297%3A16614) - Rohan will share link on slack

**Comment by rohan-arthur** [comment]
Meeting Notes:

- [ ] 1. UQI components client side implementation will be done by next week - Ayush
         - [ ] a. 2/4 done - waiting for Hetu's approval 
- [ ] 2. Validations might need some more inputs, Ayush to talk to Rashmi about this and get back - Rashmi
         - [ ] a. Clarify problem statement with @rohan-arthur 
         - [ ] b. Break down into smaller problems
         - [ ] c. Pair with others for each part
- [ ] 3. Form to JSON conversion component for plugins . - Ayush/Aman
         - [ ] a. Data migration, by tomorrow
         - [ ] b. Server side migrations (for .data), will be done by next week - @nidhi-nair 
- [ x] 4. ADS review - Aman/Ayush
- [ ] 5. Internal documentation for the components (for the use of backend) - Ayush
- [ x] 6. Firestore plugin changes should be merged this week
- [ ] 7. MongoDB form to native query conversion. Raised PR, one round of review done. - Sumit
        - [ ] a. [PoC] convert form to native and read event until user tries to edit the native query
- [ ] 8. S3 PR cypress test failures needs to be looked into - Sumit
- [ ] 9. Airtable is ready for QA again, this should be on prod by next week - Nidhi
- [ ] 10. Saas app is currently bug free. We'll be copying this app into release and connecting the current one to prod. Deployable version will be ready after the git sync issue is fixed) - Nidhi
- [ ] 11. Decide on design for request body components - Rohan will share link on slack
- [ ] 12. Google sheets component implementation will start as the components get implemented. The migration is blocked by adding support for native and appsmith modes - Nidhi
- [ ] 13. Need to create Github issues for all CRUD operations that are missing for plugins
- [x ] 14. Need to take a look at whether the component implementation by next week will take care of all CRUD operations
- [ ] 15. Need to create issues to track out of scope items for later: Field wise error handling
- [ x] 16. PR to add server side support for Pagination, Sorting and Projection is on review cycle and should be closed by this week end
- [ ] 17. Add pagination, sorting, projection components to plugins - Sumit
        - [ ] a. blocked on client side components. Pagination and sorting done already.

**Comment by rohan-arthur** [comment]
Meeting Notes:

- [ ] 1. UQI components client side implementation will be done by next week - Ayush
         - [ ] a. 2/4 done - waiting for Apeksha's review - blocked on dynamic trigger functionality
         - [ ] entity selector
         - [ ] projection selector
- [ ] 2. Form to JSON conversion component for plugins . - Ayush/Aman
         - [ ] a. Data migration - 3 story points remaning
         - [ ] b. Server side migrations (for .data), PR raised - @nidhi-nair 
         - [ ] c. Change the way we consume data in the plugins - @nidhi-nair 2 story points
- [x] 3. ADS review - Aman/Ayush
- [ ] 4. Internal documentation for the components (for the use of backend) - Ayush
- [x] 5. Firestore plugin changes should be merged this week
- [ ] 6. MongoDB form to native query conversion. Raised PR, one round of review done. - Sumit
        - [ ] a. [PoC] convert form to native and read event until user tries to edit the native query - @AmanAgarwal041  
- [ ] 7. S3 PR cypress test failures needs to be looked into, changes needed in cypress - Sumit/Ade
- [ x] 8. Airtable is ready for QA again, this should be on prod by next week - Nidhi
- [x ] 9. Saas app is currently bug free. We'll be copying this app into release and connecting the current one to prod. Deployable version will be ready after the git sync issue is fixed) - Nidhi
- [ ] 10. Decide on design for request body components - Decide on this
- [ ] 11. Google sheets component implementation will start as the components get implemented. The migration is blocked by adding support for native and appsmith modes - Nidhi
- [ ] 12. Need to create Github issues for all CRUD operations that are missing for plugins
- [x] 13. Need to take a look at whether the component implementation by next week will take care of all CRUD operations
- [ ] 14. Need to create issues to track out of scope items for later: Field wise error handling
- [x] 15. PR to add server side support for Pagination, Sorting and Projection is on review cycle and should be closed by this week end
- [ ] 16. Add pagination, sorting, projection components to plugins - Sumit
        - [ ] a. blocked on client side components. Pagination and sorting done already.
- [ ] 17. Validations #5702 
          - [ ] a. Extracting requirements from JSON #10230
          - [ ] b. Supply config to validations service #10231
          - [ ] c. Implementing types of validations #10232
          - [ ] d. Show errors #10233
          - [ ] e. DynamicTextControl and DynamicInputTextControl #10234
          - [ ] f. Backend Changes and Migration #10235

**Comment by rohan-arthur** [comment]
Meeting Notes:

- [x] 1. UQI components client side implementation will be done by next week - Ayush
         - [ x] done
         - [ x] entity selector
         - [ x] projection selector
- [ ] 2. Form to JSON conversion component for plugins . - Ayush/Aman
         - [ ] a. Data migration - 3 story points remaning
         - [ ] b. Server side migrations (for .data), PR raised - @nidhi-nair 
         - [ ] c. Change the way we consume data in the plugins - @nidhi-nair 2 story points
- [x] 3. ADS review - Aman/Ayush
- [ ] 4. Internal documentation for the components (for the use of backend) - Ayush
      - [ ] https://github.com/appsmithorg/uqi-json-rules - needs review
- [x] 5. Firestore plugin changes should be merged this week
- [ ] 6. MongoDB form to native query conversion. Raised PR, one round of review done. - Sumit
        - [ ] a. [PoC] convert form to native and read event until user tries to edit the native query - @Irongade 
        - [ ] b. decision in brief: we should keep allowing the user to switch between form and raw mode, and every form change should be reflected in raw mode every time, UNTIL the user *attempts to* make any change in the raw mode - from that point onwards, no more changes will be reflected on either side.  
        - [ x] c. should we resume conversion support in case the user completely clears out all their entry in the raw mode? (remember we don't support raw to form parsing) - can consider this as an additional layer of conditional logic, depending on user feedback.
        - [x ] d. another option is to use a button and let the user decide when they want to let the conversion consume - but this is complicated by the fact that the conversion happens not only in the find but in several other elements.
- [x ] 7. S3 PR cypress test failures needs to be looked into, changes needed in cypress - Sumit/Ade
- [ x] 8. Airtable is ready for QA again, this should be on prod by next week - Nidhi
- [x ] 9. Saas app is currently bug free. We'll be copying this app into release and connecting the current one to prod. Deployable version will be ready after the git sync issue is fixed) - Nidhi
- [ ] 10. Decide on design for request body components - Decide on this
- [ ] 11. Google sheets component implementation will start as the components get implemented. The migration is blocked by adding support for native and appsmith modes - Nidhi
- [ ] 12. Need to create Github issues for all CRUD operations that are missing for plugins
- [x] 13. Need to take a look at whether the component implementation by next week will take care of all CRUD operations
- [ ] 14. Need to create issues to track out of scope items for later: Field wise error handling
- [x] 15. PR to add server side support for Pagination, Sorting and Projection is on review cycle and should be closed by this week end
- [ ] 16. Add pagination, sorting, projection components to plugins - Sumit
        - [ ] a. blocked on client side components. Pagination and sorting done already.
- [ ] 17. Validations #5702 
          - [ ] a. Extracting requirements from JSON #10230 - review
          - [ ] b. Supply config to validations service #10231 - review
          - [ ] c. Implementing types of validations #10232 - review
          - [x ] d. Show errors #10233
          - [ ] e. DynamicTextControl and DynamicInputTextControl #10234  - pending
          - [ ] f. Backend Changes and Migration #10235 - pending

**Comment by nidhi-nair** [comment]
Meeting Notes:

 1. UQI components available on release, consumption required on server side
 - [ ] Waiting on projection component changes @ayushpahwa by EOD
 - [ ] Pagination and sorting for S3 is in progress @sumitsum 
 2. Form to JSON conversion component for plugins 
- [ ] a. Data consumption on client side - @ayushpahwa 
- [x] b. Server side migrations (for .data), PR raised - @nidhi-nair
- [ ] c. Change the way we consume data in the plugins - @nidhi-nair 2 story points
 3. Internal documentation for the components (for the use of backend) - Ayush
- [ ] https://github.com/appsmithorg/uqi-json-rules - needs review
 4. MongoDB form to native query conversion. Raised PR, one round of review done. - Sumit
- [ ] a. [PoC] convert form to native and read event until user tries to edit the native query - @Irongade
- [ ] b. decision in brief: we should keep allowing the user to switch between form and raw mode, and every form change should be reflected in raw mode every time, UNTIL the user attempts to make any change in the raw mode - from that point onwards, no more changes will be reflected on either side.
 5. Need to create Github issues for all CRUD operations that are missing for plugins - @nidhi-nair 
 6. Need to create issues to track out of scope items for later: Field wise error handling - @nidhi-nair 
 7. Validations Add validations for each form element component #5702
- [ ] a. Extracting requirements from JSON [Feature]: [validations] Extracting requirements from JSON #10230 - review (QA)
- [ ] b. Supply config to validations service : #10231 - review (QA)
- [ ] c. Implementing types of validations : #10232 - review (QA)
- [ ] d. Validate and show errored out action fields: #2675 - pending (Rashmi)
- [ ] e. Backend Changes and Migration: #10235 - pending (not to be picked up this week)
8. Postgres migration - @sumitsum

**Comment by nidhi-nair** [comment]
Meeting notes:

- [ ] Mock form JSON for UX demo (S3 and Google sheets) @nidhi-nair @sumitsum 
        - Delete many - S3
        - Delete many - Google sheets
- [ ] Mark MSSQL as not supporting CRUD @nidhi-nair 
- [ ] Raw data for each plugin TBD
- [ ] Memoization related bug for components @ayushpahwa 
- [ ] Pagination and sorting for S3 in review @sumitsum 
- [ ] Waiting on projection component changes @ayushpahwa
- [ ] Data consumption on client side for JSON/Form mode @ayushpahwa 
- [ ] UQI JSON rules PR to be created @ayushpahwa 
- [ ] Convert form to native and read event until user tries to edit the native query  - Blocked on form/JSON mode @Irongade 
- [ ] 1 validation PR merged, check if this can be consumed @rashmi-rai 
- [ ] Postgres migration @sumitsum 
- [ ] Mock form JSON for UX demo (Firestore, Mongo, SQL) @nidhi-nair @sumitsum 
        - Bulk operations - Firestore

**Comment by rohan-arthur** [comment]
V1 scope: https://www.notion.so/appsmith/Sequencing-Tasks-for-UQI-epic-831ed68a652140c696e9535300673860 

1. [Implementing components](https://app.zenhub.com/workspaces/be-coders-pod-6167d73fab4b07001219a3d0/board?epics=Z2lkOi8vcmFwdG9yL0VwaWMvNzEzNzc0&labels=uqi%20components) in ~~datasource and~~ query forms, that can be used by the app developer
    1. Timeline - 23 Feb 2022
    2. form to json switch
    3. projection component
- [ ] #10762 
- [ ] #10761 
- [ ] #9656 
- [ ] #10245 
- [ ] #11160 - discussion needed between @rashmi-rai @ayushpahwa @ApekshaBhosale 
- [ ] #2809 
- [ ] #10109 

    
2. Documenting components so that the plugin developer can more easily create plugins
    1. Timeline?

3. Migrating existing plugins to new component based format - 4 (S3, Firestore, Mongo in progress. SQL and GSheets remaining, out of which we may not do SQL)
    1. S3, Firestore, Mongo - timeline is 2 weeks
    2. GSheets - timeline is 1 week
    3. SQL - descope

**Comment by rohan-arthur** [comment]
V1 scope: https://www.notion.so/appsmith/Sequencing-Tasks-for-UQI-epic-831ed68a652140c696e9535300673860 

1. [Implementing components](https://app.zenhub.com/workspaces/be-coders-pod-6167d73fab4b07001219a3d0/board?epics=Z2lkOi8vcmFwdG9yL0VwaWMvNzEzNzc0&labels=uqi%20components) in ~~datasource and~~ query forms, that can be used by the app developer
    1. Timeline - 23 Feb 2022
    2. form to json switch
    3. projection component
- [ ] #10762 in review @ayushpahwa
- [ ] #10761 in review @ayushpahwa
- [ ] #10245 @ayushpahwa
      - [ ] migrate sample data for 3 actions @nidhi-nair , @ayushpahwa 
- [ ] #9656 @ayushpahwa
      - [ ] server side migration @nidhi-nair 
- [ ] #11160 - in review, some issues with entity selector 
- [ ] #2809 
- [ ] #10109 

    
2. Documenting components so that the plugin developer can more easily create plugins
    1. Timeline?

3. Migrating existing plugins to new component based format - 4 (S3, Firestore, Mongo in progress. SQL and GSheets remaining, out of which we may not do SQL)
    1. S3, Firestore, Mongo - timeline is 2 weeks
    2. GSheets - timeline is 1 week
    3. SQL - descope

**Comment by rohan-arthur** [comment]
V1 scope: https://www.notion.so/appsmith/Sequencing-Tasks-for-UQI-epic-831ed68a652140c696e9535300673860 

1. [Implementing components](https://app.zenhub.com/workspaces/be-coders-pod-6167d73fab4b07001219a3d0/board?epics=Z2lkOi8vcmFwdG9yL0VwaWMvNzEzNzc0&labels=uqi%20components) in ~~datasource and~~ query forms, that can be used by the app developer
    1. Timeline - 28 Feb 2022
    2. form to json switch
    3. projection component
- [x ] #10762 in review @ayushpahwa
- [ x] #10761 in review @ayushpahwa
- [ ] #10245 @ayushpahwa in review 
      - [x ] migrate sample data for 3 actions @nidhi-nair , @ayushpahwa 
      - [ x] create test instance for migration - @nidhi-nair 
      - [ x] Get test cases from @prapullac - @ayushpahwa 
- [ ] #9656 @ayushpahwa in review
      - [ ] server side migration @nidhi-nair  - waiting on the client-side PR
      - [ x] merge into 10245
- [ x] #11464 - in review
- [x ] #11160 - evaluation order 
- [ ] #2809 
- [ ] #10109 
- [ ] [ui fixes](https://theappsmith.slack.com/archives/C02KTH6EHUG/p1646115609033869) - #9302 
- [ ] Share usage of all help text and other meta props for UQI forms - @ayushpahwa [in process link](https://www.notion.so/appsmith/How-to-write-a-new-plugin-5115d4db32f5483982ae6e586eb09a27)

    
2. Documenting components so that the plugin developer can more easily create plugins
    1. Timeline?

3. Migrating existing plugins to new component based format - 4 (S3, Firestore, Mongo in progress. SQL and GSheets remaining, out of which we may not do SQL)
- [ ] GSheets 
    - [ ] where clause and existing features - with the new ux #6591
    - [ ] other components sorting, projection etc #9519 
- [ ] S3
- [ ] Firestore, Mongo
(SQL - descope)

**Comment by prapullac** [comment]
Identified Issues :
1)  When a Query has a drop-down and it is converted to JSON format it is observed that the data remains to what is written in the JSON and does not get updated wrt selection option of the dropdown menu.
2) https://github.com/appsmithorg/appsmith/issues/11871
